es/simplifyCostsDocument.go

```diff
@@ -0,0 +1,189 @@
+		scd.ChildrenKind: children,
+	}
+}
+
+type value = map[string]interface{}
+type aggregation = map[string]interface{} // aggregation has buckets
+type bucket = map[string]interface{}      // bucket has aggregations and values
+
+const (
+	BucketPrefix         = "by-"
+	BucketKeyKey         = "key"
+	AggBucketKey         = "buckets"
+	BucketKeyAsStringKey = "key_as_string"
+	BucketValueKey       = "value"
+	BucketValueValueKey  = "value"
+)
+
+var (
+	ErrNoSingleRootAggregation   = errors.New("document does not have a single aggregation at its root")
+	ErrNoSingleAggregationBranch = errors.New("document's aggregations branch")
+	ErrFailedJsonParsing         = errors.New("failed to parse JSON document")
+	ErrKeyNotFound               = errors.New("could not find 'key' field")
+	ErrNoAggregation             = errors.New("found no next aggregation and no value")
+)
+
+func SimplifyCostsDocument(ctx context.Context, sr *elastic.SearchResult) (SimplifiedCostsDocument, error) {
+	var scdz SimplifiedCostsDocument
+	if len(sr.Aggregations) == 1 {
+		for k, v := range sr.Aggregations {
+			if v != nil {
+				return simplifyCostsDocumentWithSingleAggregation(ctx, k, sr.Aggregations[k])
+			}
+		}
+	}
+	return scdz, ErrNoSingleRootAggregation
+}
+
+func simplifyCostsDocumentWithSingleAggregation(ctx context.Context, rootAgg string, rm *json.RawMessage) (SimplifiedCostsDocument, error) {
+	var logger = jsonlog.LoggerFromContextOrDefault(ctx)
+	var parsedDocument bucket
+	var scdz SimplifiedCostsDocument
+	err := json.Unmarshal(*rm, &parsedDocument)
+	if err != nil {
+		logger.Error("Failed to parse JSON costs document.", err.Error())
+		return scdz, ErrFailedJsonParsing
+	} else {
+		return simplifyCostsDocumentRec(ctx, map[string]interface{}{rootAgg: parsedDocument}, true)
+	}
+}
+
+func simplifyCostsDocumentRec(ctx context.Context, doc bucket, root bool) (SimplifiedCostsDocument, error) {
+	var scd SimplifiedCostsDocument
+	if !root {
+		var err error
+		scd.Key, err = getKey(doc)
+		if err != nil {
+			return scd, err
+		}
+	}
+	if value, ok := getValue(ctx, doc); ok {
+		scd.HasValue = true
+		scd.Value = value
+		return scd, nil
+	} else if childrenKind, children, err := getChildren(ctx, doc); err == nil {
+		scd.Children = children
+		scd.ChildrenKind = childrenKind
+		return scd, nil
+	} else {
+		return scd, err
+	}
+}
+
+func getKey(doc bucket) (string, error) {
+	if key, ok := doc[BucketKeyKey]; ok {
+		if tkey, ok := key.(string); ok {
+			return tkey, nil
+		}
+	}
+	if key, ok := doc[BucketKeyAsStringKey]; ok {
+		if tkey, ok := key.(string); ok {
+			return tkey, nil
+		}
+	}
+	return "", ErrKeyNotFound
+}
+
+func getValue(ctx context.Context, doc bucket) (float64, bool) {
+	var logger = jsonlog.LoggerFromContextOrDefault(ctx)
+	if value, ok := doc[BucketValueKey]; ok {
+		if tvalue, ok := value.(map[string]interface{}); ok {
+			if valuevalue, ok := tvalue[BucketValueValueKey]; ok {
+				if tvaluevalue, ok := valuevalue.(float64); ok {
```

Comment:
`value` then `tvalue` then `valuevalue` and finally `tvaluevalue`, you could name them better here